### ISSUE TITLE
Error: Invalid JSON data. If data is mysql JSON field

### ISSUE THREAD

**Issue by d3artagnan** [dmitry-kulikov/yii2-json-editor#5]
### What steps will reproduce the problem?
- Load data from DB via ActiveRecord Model. One of the fields has JSON type, so starting from 2.0.14 it is converted to Array while populating
- Use Active form and use the widget, for example:
```
<?= $form->field($model, 'test'
        ->widget(JsonEditor::class, [
            'clientOptions' => ['modes' => ['code', 'tree']],
        ]);
    ?>
```

$model->test will contains array, which is a default behavior when retrieved JSON data type  from db, yii do something like
`$placeholder = $this->queryBuilder->bindParam(Json::encode($value), $params);`

### What is the expected result?
widget can read data array from JSON data type

### What do you get instead?
Error: Invalid JSON data.

### Additional info
Please check at JsonEditor.php
`Line: 170`
`$encodedValue = Json::htmlEncode(Json::decode($this->value, false));`

Maybe we can add a value checking first before decode the value:
`$encodedValue => is_array($this->value) ? Json::htmlEncode($this->value) : Json::htmlEncode(Json::decode($this->value, false));`

| Q                   | A 
|---------------------|---
| Extension version   | 2.2.0
| Yii version         | 2.0.27
| JSON editor version | 
| PHP version         | 5.3
| HHVM version        | 
| Operating system    |

**Comment by dmitry-kulikov** [comment]
Thanks for the report, indeed it is bug. Unfortunately proposed fix will not work for JSON strings. There may be string `"foo"` in database, it is valid JSON. Yii will automatically convert it to `foo` - it is string, but not JSON anymore, so same error will occur. I'll look for better solution at weekend.

**Comment by d3artagnan** [comment]
I have 2 approach for this scenario,

### 1st approach
 is to decode the JSON field in the model to string from array; in view or controller beforehand by doing

`$model->field = Json::decode($model->field);`

this way, we wont need to change anything in the yii2-json-editor widget

### 2nd approach
is to make the yii2-json-editor widget to handle array from JSON field at `JsonEditor.php`
`Line: 170`
```
$encodedValue = is_array($this->value) ? Json::htmlEncode($this->value) :
            Json::htmlEncode(Json::decode($this->value, false));
$this->options['value'] = isset($this->options['value']) ? $this->options['value'] :
            (is_array($this->value) ? Json::encode($this->value) : $this->value);
```
this can handle both json and string like `"foo"` from database.
and the `$this->options['value']` code is use to feed Yii Active Input, since it can only received a string.

### Either way, 
we still need to convert the POST field back to array in controller, to save the data back to the valid JSON database field, and not a regular string, by doing
```
if ($model->load(\Yii::$app->request->post())) {
            $model->field = Json::decode($model->field);
```
Since Yii will return a string not an array from Active Input,

What do you think?

**Comment by dmitry-kulikov** [comment]
I suggest to use 1st approach for now:
```php
$model->field = Json::encode($model->field);
```
2nd approach could work, but it is not so simple. It will not work for JSON `"foo"`, because it is `"foo"` in database, but then Yii automatically decodes this JSON and it will turn to `foo` (without double quotes), this is not array and `Json::decode` will be executed, but it is not JSON anymore, thus error will happen. We could catch error, like "if we cannot decode it then it is not JSON" and build logic based on it. Or use regular expression or something else. Still I think attempt to detect JSON is wrong. It will not give significant benefits but may produce bugs. I'm going to add `decodedValue`:
```php
$model->data = ['foo' => 'bar'];
echo $form->field($model, 'data')
    ->widget(JsonEditor::class, ['decodedValue' => $model->data]);
```
If you need to pass already decoded JSON then just let widget know it is decoded.
Actually I already did it, but would like to add tests.

Regarding controller and ActiveInput. I think you will have to decode manually or override method `load`:
```php
public function load($data, $formName = null)
{
    if (!parent::load($data, $formName)) {
        return false;
    }

    $this->data = Json::decode($this->data, false);

    return true;
}
```

**Comment by dmitry-kulikov** [comment]
I released 2.3.0, it adds property `decodedValue`, which works as I described above. Also I updated documentation to make it less ambiguous. I'm going to close this issue since usage of `decodedValue` is simple enough and documented. If you disagree then you can override `JsonEditor::determineValue()` and implement behavior which you like.
